### ISSUE TITLE
Is it possible to get available cypher variables ?

### ISSUE THREAD

**Issue by K-Lovelace** [neo4j-contrib/cypher-dsl#204]
Hi,

Would it be possible, given an ongoing query having multiple matches and with, to know what cypher variables are availables ? The idea would be to add to a with every variables present plus some other expression we need to calculate (for example with a case when).

So given a cypherdsl code generating the unfinished query:

```
MATCH (a:Label)--(b)
WITH a, b
OPTIONAL MATCH (b)-->(c)
WITH a, c, CASE WHEN c IS NULL OR c:Label THEN TRUE ELSE FALSE END AS isNullOrLabel
```

I would like to do something like : 

```
List<Expression> availableExpressions = query.availableExpressions();
```

and have as a result in the list : `a, c, isNullOrLabel`. (b is not in the last with so we do not want it in this list, it's no longer available)

For now, in our query generator, we keep track of what's available using a custom map, but it would be much cleaner if it was part of the API. 

The same thing with already set params (Parameter objects) would be awesome too.

Is any of this imaginable ?

**Comment by michael-simons** [comment]
So to get this right: You want to have only those expression that are returned or in the last step of a partial query?
I see how this is being useful, would also be nice in the `ExecutableResultStatement`. I think about it.

**Comment by K-Lovelace** [comment]
Yes exactly, doing this on our own is quite a headache. I look forward to your insight on this :+1:

**Comment by michael-simons** [comment]
Hi @K-Lovelace you write 

> The same thing with already set params (Parameter objects)

which is not accurate: Those information are only available when the statement is finally build, not during an unfinished query.
For your purpose I guess you want these information as you go, available in the unfinished statement as well?

**Comment by K-Lovelace** [comment]
Hi @michael-simons, 

Yes exactly, query.availableParameters() returning a List of Paramater objects, added to the query through conditions, node properties and so on. We actually keep track of them on our own for now, in order for example to reuse a parameter value from the start of the query instead of adding another one with a different name but the same value.

Is it clearer for you ? 

(I guess I could do without this, it's easier to keep track of what parameters have been added compared to what variables are available in a "cypher" way.)

**Comment by michael-simons** [comment]
> Is it clearer for you ?

No, not really. 
Is it enough to have the available expressions after the query has been finished building or is that too late?

**Comment by K-Lovelace** [comment]
What I really want is a way, during the construction of the query, to have for example an OngoingReading::availableExpressions() method that returns the list of currently available expressions (a, c, isNullOrLabel in my previous example) so that I can, for example, add them in a new with clause along with additional informations, without having to keep track myself of the variables. 

I also need the same method for parameters, giving the parameters that are already present in the query, meaning the parameter added in a isEqualTo or withProperties for example.

> Is it enough to have the available expressions after the query has been finished building or is that too late?

Yes, it would be too late, I need them during the query construction.

**Comment by michael-simons** [comment]
I have added API on the ongoing `WITH` and `RETURN` and on the statement as a whole.
Please let me know if this works for you. I don't see many other extension points to do this.

Please look here for usage atm: https://github.com/neo4j-contrib/cypher-dsl/blob/main/neo4j-cypher-dsl/src/test/java/org/neo4j/cypherdsl/core/IdentifiableExpressionsIT.java

**Comment by michael-simons** [comment]
This has been released as [2021.3.2](https://github.com/neo4j-contrib/cypher-dsl/releases/tag/2021.3.2) in case you wanna try it.
